src/components/product-switcher/index.tsx

```diff
@@ -0,0 +1,243 @@
+  const shouldFocusFirstAnchor = useRef<boolean>(false)
+
+  useLayoutEffect(() => {
+    if (!isOpen) {
+      return
+    }
+
+    // Focuses the first anchor element if needed
+    if (shouldFocusFirstAnchor.current) {
+      const firstAnchorElement = getAllAnchorElements()[0]
+      firstAnchorElement.focus()
+      shouldFocusFirstAnchor.current = false
+    }
+
+    const handleDocumentClick = (e) => {
+      const isClickOutside = !productChooserRef.current.contains(e.target)
+      if (isClickOutside) {
+        setIsOpen(false)
+      }
+    }
+
+    document.addEventListener('click', handleDocumentClick)
+
+    return () => document.removeEventListener('click', handleDocumentClick)
+  }, [isOpen])
+
+  /**
+   * It would be more optimal to set onKeyDown for the containing element, but that is not allowed
+   * on a <div> with no role. We do not want to use a menu role (see https://adrianroselli.com/2017/10/dont-use-aria-menu-roles-for-site-nav.html)
+   * so there is no role currently set on the containing <div>. If we find an appropriate role, then we can
+   * change the approach. For now, it's most appropriate to set onKeyDown on the <button> and <ul>,
+   * hence the `KeyboardEventHandler<HTMLButtonElement | HTMLUListElement>` definition.
+   */
+  const handleKeyDown: KeyboardEventHandler<
+    HTMLButtonElement | HTMLUListElement
+  > = (e) => {
+    const isEscapeKey = e.key === 'Escape'
+    const isEnterKey = e.key === 'Enter'
+    const isSpaceKey = e.key === ' '
+
+    /**
+     * The reason we can't focus the first anchor here is because we have the anchor element
+     * is not rendered until after `setIsOpen(true)` is called by the <button>'s onClick (default
+     * behavior of <button> elements).
+     *
+     * Might be possible to do e.preventDefault and then do the focus() call here? Not sure we
+     * need to do that though unless we are very against this approach. I'd rather not prevent
+     * default behavior if we don't have to.
+     */
+    if (!isOpen && (isEnterKey || isSpaceKey)) {
+      shouldFocusFirstAnchor.current = true
+      return
+    }
+
+    if (isEscapeKey) {
+      setIsOpen(false)
+      buttonRef?.current?.focus()
+      return
+    }
+  }
+
+  const renderProductListItem = (
+    product: Product,
+    index: number
+  ): ReactElement => {
+    const isCurrentProduct = product.code === currentProduct.code
+
+    const handleAnchorKeyDown: KeyboardEventHandler<HTMLAnchorElement> = (
+      e
+    ) => {
+      const lastIndex = products.length - 1
+      const isFirstItem = index === 0
+      const isLastItem = index === lastIndex
+      if (!(isFirstItem || isLastItem)) {
+        return
+      }
+
+      const isTabbingForward = !e.shiftKey && e.key === 'Tab'
+      const isTabbingBackward = e.shiftKey && e.key === 'Tab'
+      if (!(isTabbingForward || isTabbingBackward)) {
+        return
+      }
+
+      const anchorElements = getAllAnchorElements()
```

Comment:
It looks like we are only concerned with focusing the first and last elements in the list. We could use two static refs for this, and that way we wouldn't need to query the DOM when we do need to put focus elsewhere.

```tsx
const firstAnchorRef = useRef()
const lastAnchorRef = useRef()

// ...

let ref = null
if (isFirstItem) ref = firstAnchorRef
if (isLastItem) ref = lastAnchorRef

// ...

<a
  ref={ref}
  // ...
```

Reply:
Helpful hook for managing active indexes that wrap https://github.com/souporserious/use-roving-index

Reply:
This is such a great idea! 🙌🏻  I'll work on making this change today.

Reply:
Ha, @alexcarpenter and I replied at almost the same time. 😆  My previous comment was directed at @BRKalow, but I will definitely take a look at that hook Alex! I've not seen it before. :til:

Reply:
@ashleemboyer ha no problem, just sharing as a reference. Managing indexes can be common, so a shared hook might be a helpful abstraction. No worries if not.

Reply:
Well, since I've added groups within the the list now, I don't think the hook is going to help us anymore. But out of curiosity and to make sure I'm understanding it correctly, it seems like it's most useful when we have to manage the focus of items that don't have default handling? In this instance I just want to override the handling of two items and don't have to manage how focus works for all of them.

Reply:
Also, I've updated the code to work with 2 refs now instead of querying the document for anchors!